src/Middleware/HeaderPropagation/src/DependencyInjection/HeaderPropagationExtensions.cs

```diff
@@ -0,0 +1,95 @@
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    public static class HeaderPropagationExtensions
+    {
+        private static readonly string _unableToFindServices = string.Format(
+            "Unable to find the required services. Please add all the required services by calling '{0}.{1}' inside the call to 'ConfigureServices(...)' in the application startup code.",
+            nameof(IServiceCollection),
+            nameof(AddHeaderPropagation));
+
+        /// <summary>
+        /// Adds services required for propagating headers to a <see cref="HttpClient"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static IServiceCollection AddHeaderPropagation(this IServiceCollection services)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            services.TryAddSingleton<HeaderPropagationValues>();
+
+            return services;
+        }
+
+        /// <summary>
+        /// Adds services required for propagating headers to a <see cref="HttpClient"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <param name="configureOptions">The <see cref="HeaderPropagationOptions"/> to configure the middleware with.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static IServiceCollection AddHeaderPropagation(this IServiceCollection services, Action<HeaderPropagationOptions> configureOptions)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configureOptions == null)
+            {
+                throw new ArgumentNullException(nameof(configureOptions));
+            }
+
+            services.Configure(configureOptions);
+            services.AddHeaderPropagation();
+
+            return services;
+        }
+
+        /// <summary>
+        /// Adds a message handler for propagating headers collected by the <see cref="HeaderPropagationMiddleware"/> to a outgoing request.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHttpClientBuilder"/> to add the message handler to.</param>
+        /// <returns>The <see cref="IHttpClientBuilder"/> so that additional calls can be chained.</returns>
+        public static IHttpClientBuilder AddHeaderPropagation(this IHttpClientBuilder builder)
+        {
+            builder.Services.TryAddSingleton<HeaderPropagationValues>();
+            builder.Services.TryAddTransient<HeaderPropagationMessageHandler>();
```

Comment:
I think this could be moved to `services.AddHeaderPropagation` - it is a service after all, and It can be used without the factory.

Reply:
ok